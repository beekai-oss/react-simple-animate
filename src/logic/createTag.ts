import createStyle from './createStyle';
import { Keyframes } from '../types';

export default function createTag({
  keyframes,
  animationName,
}: {
  keyframes: Keyframes;
  animationName: string;
}): {
  styleTag: any;
  index: number;
} {
  let styleTag = document.querySelector('style[data-id=rsi]');
  let index = 0;

  if (!styleTag) {
    styleTag = document.createElement('style');
    styleTag.setAttribute('data-id', 'rsi');
    document.head.appendChild(styleTag);
  }

  try {
    // @ts-ignore
    index = styleTag.sheet.cssRules.length;
  } catch (e) {
    index = 0;
  }

  try {
    // @ts-ignore
    styleTag.sheet.insertRule(
      createStyle({
        keyframes,
        animationName,
      }),
      index,
    );
  } catch (e) {
    console.error('react simple animate, error found during insert style ', e); // eslint-disable-line no-console
  }

  return {
    styleTag,
    index,
  };
}
