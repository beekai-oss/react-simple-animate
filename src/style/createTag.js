import createStyle from './createStyle';
import createRandomName from '../utils/createRandomName';

export default function createTag(keyframes) {
  const animationName = createRandomName();
  let styleTag = document.querySelector('style[data-id-rsi]');

  if (!styleTag) {
    styleTag = document.createElement('style');
    styleTag.setAttribute('data-id', 'rsi');
    document.head.appendChild(styleTag);
  }

  const index = styleTag.sheet.cssRules.length;

  try {
    this.styleTag.sheet.insertRule(
      createStyle({
        keyframes,
        animationName,
      }),
      index,
    );
  } catch (e) {
    console.error('react simple animate, error found during insert style:', e); // eslint-disable-line no-console
  }

  return {
    styleTag,
    index,
    animationName,
  };
}
