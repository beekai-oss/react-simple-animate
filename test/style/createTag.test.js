import createTag from '../../src/style/createTag';

jest.mock('../../src/style/createStyle', () => () => 'test');

const documentCreateOriginal = document.createElement;
const appendChildOriginal = document.head.appendChild;

describe('createTag', () => {
  beforeEach(() => {
    document.createElement = jest.fn();
    document.head.appendChild = jest.fn();
  });

  afterEach(() => {
    document.createElement = documentCreateOriginal;
    document.head.appendChild = appendChildOriginal;
  });

  it('should the tag and total length of css rules', () => {
    const insertRule = jest.fn();
    document.createElement.mockReturnValue({ sheet: { cssRules: [1, 2], insertRule }, setAttribute: () => {} });
    const output = createTag({ keyframes: '', animationName: 'test' });

    expect(insertRule).toBeCalled();
    expect(output).toMatchSnapshot();
  });

  it('should still return tag and index even insertRule thow error', () => {
    const insertRule = () => {
      throw new Error('failed');
    };
    document.createElement.mockReturnValue({ sheet: { cssRules: [1, 2], insertRule }, setAttribute: () => {} });
    const output = createTag({ keyframes: '', animationName: 'test' });

    expect(output).toMatchSnapshot();
  });
});
