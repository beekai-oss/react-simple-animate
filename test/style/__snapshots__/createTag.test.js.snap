// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createTag should still return tag and index even insertRule thow error 1`] = `
Object {
  "index": 2,
  "styleTag": Object {
    "setAttribute": [Function],
    "sheet": Object {
      "cssRules": Array [
        1,
        2,
      ],
      "insertRule": [Function],
    },
  },
}
`;

exports[`createTag should the tag and total length of css rules 1`] = `
Object {
  "index": 2,
  "styleTag": Object {
    "setAttribute": [Function],
    "sheet": Object {
      "cssRules": Array [
        1,
        2,
      ],
      "insertRule": [MockFunction] {
        "calls": Array [
          Array [
            "test",
            2,
          ],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      },
    },
  },
}
`;
